### PR TITLE
SI-5788 Tailcalls LabelDefs correctly duplicated

### PR DESCRIPTION
**Pull request by VladUreche**
... in specialization.

This is a quick hack to get SI-5788 fixed in 2.10.x. The full patch,
which fixes the tailcalls LabelDefs will be merged into trunk, as it's
too late for big changes.

For reference, the complete fix is: e86afe65c8

Review by @adriaanm or @non.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Duplicators.scala [hunk 1]
```diff
@@ -41,7 +41,7 @@ abstract class Duplicators extends Analyzer {
   }
 
   protected def newBodyDuplicator(context: Context) = new BodyDuplicator(context)
-  
+
```
Comment (jsuereth) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:44]: But _this_ blank line is BETTER!

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:44]: Hi Josh, sorry, that's Sublime automatically killing trailing whitespaces. But I think it actually removed whitespace from that line, as you can see it both removed and then added.

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:44]: I've been doing the same thing.  Soon as we a) enforce no trailing whitespace is checked in and b) do one more remove-all-trailing-whitespace commit then the invisible line dancing will come to a close.

File: src/compiler/scala/tools/nsc/typechecker/Duplicators.scala [hunk 2]
```diff
@@ -278,12 +278,19 @@ abstract class Duplicators extends Analyzer {
...
+          // to do it manually here -- but for the tailcalls-generated labels, ValDefs are created before the LabelDef,
+          // so we just need to plug in the name
+          def newParam(p: Tree): Ident =
+            if (isTailLabel)
+              Ident(p.symbol.name) // let the typer pick up the right symbol
+            else {
+              val newsym = p.symbol.cloneSymbol //(context.owner) // TODO owner?
+              Ident(newsym.setInfo(fixType(p.symbol.info)))
+            }
           val params1 = params map newParam
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:294]: could this also work?

```
val params1 = if (isTailLabel) params else params map newParam
```

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:294]: The trick is that the symbols are already duplicated when they are encountered in ValDefs, so it's not the LabelDef's parameter symbols I'm after, but their equivalents after the ValDef duplication. That's why I only give the names and let the typer find them.

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/typechecker/Duplicators.scala:294]: A mapping from the original symbols to the new ones would also do the job.

**Comment by VladUreche** [comment]
@adriaanm, is this GTG?

**Comment by adriaanm** [comment]
I guess it would be more regular to use the right symbol to create the Ident, but I'm okay either way
this whole LabelDef business is one big hack anyway

LGTM

**Pull request merged**